programs/bpf_loader/src/serialization.rs

```diff
@@ -437,24 +437,26 @@ fn serialize_parameters_aligned(
                 size += size_of::<u8>() // is_signer
                 + size_of::<u8>() // is_writable
                 + size_of::<u8>() // executable
-                + size_of::<u32>() // original_data_len
+                + 4 // padding  of 4 bytes after executable
+                + size_of::<u64>() // original_data_len
                 + size_of::<Pubkey>()  // key
                 + size_of::<Pubkey>() // owner
                 + size_of::<u64>()  // lamports
                 + size_of::<u64>()  // data len
-                + MAX_PERMITTED_DATA_INCREASE
                 + size_of::<u64>(); // rent epoch
                 if copy_account_data {
-                    size += data_len + (data_len as *const u8).align_offset(BPF_ALIGN_OF_U128);
+                    let align_offset = (data_len as *const u8).align_offset(BPF_ALIGN_OF_U128);
+                    size += data_len + align_offset;
+                    size += MAX_PERMITTED_DATA_INCREASE + align_offset;
                 } else {
-                    size += BPF_ALIGN_OF_U128;
+                    size += MAX_PERMITTED_DATA_INCREASE + BPF_ALIGN_OF_U128;
                 }
             }
         }
     }
     size += size_of::<u64>() // data len
     + instruction_data.len()
-    + size_of::<Pubkey>(); // program id;
+    + size_of::<Pubkey>() + 32; // program id; reserve 32 bytes as extra space to avoid reallocation
```

Comment:
We need an explanation for the `+ 32`

Reply:
Yeah this is debatable. I added in my code because I prefer adding 32 extra bytes instead of resizing because of potential wrong calculation of bytes.

Reply:
overallocation is wrong

Reply:
Will remove it.